### ISSUE TITLE
whats the best name for the default export?

### ISSUE THREAD

**Issue by jcbhmr** [felipecrs/typedoc-plugin-rename-defaults#14]
_this is to get a feel for what others thing the default name should be so i dont get stuck in my opinion bubble_

**which of these is best?**
the project that I'm using as a demo https://isaacs.github.io/rimraf/functions/path_arg.default.html

<table><td>

"default" or "export="
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/8f8f21d0-0d7e-4c8e-a38f-233f47893cb2)

<td>

name of symbol (may be raw expr; need 2nd opt)
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/f061d145-f177-4ed3-947d-1859a008e74d)

<td>

same name as module
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/16bdcaf3-87b1-4ed0-bb7f-a90e7546354b)

<tr><td>

camelCase module .default suffix
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/0a39ee53-fbad-4bee-aa44-d6985ff0d229)

<td>

camelCased module name
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/355ddfe7-634a-49e6-9e88-a621cc37dd04)

</table>

and here's another real-ish-world example this time with dots in the module names

<table><td>

"default" or "export="
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/5355894e-8d60-40c1-bfcb-dc007f0852fb)

<td>

name of symbol (may be raw expr; need 2nd opt)
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/69e693ce-2678-43b2-97f1-ea18cc45faf8)

<td>

same name as module
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/012137de-8c90-4003-b683-a07df4c74204)

<tr><td>

camelCased module name
![image](https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/3d0f0828-9959-4dfa-b87d-ad3b97da8da3)

</table>

which ones do you think are the best? lmk if there's other ways that i haven't thought of idk

<sup>yes, this is a bit silly amount of talk/detail/work/effort/whatever just to talk about some inconsequential code naming conventions lol 🤣</sup>

**Comment by felipecrs** [comment]
What is your preferred one? It is kinda hard. I don't have a easy choice, but I am leaning towards the last option.

**Comment by jcbhmr** [comment]
after mulling over it while doing up those demos, i think i personally really like the camelCase-ized module name. why:
1. if you do `import doTheThing from "./do-the-thing.ts"` or `import otherThing from "./otherThing.ts"`, it makes sense that the default export in TypeDoc matches that `doTheThing` that you would import it as in your code
2. it works well with the example type signature (no non-JS syntax like `export=(a, b, c)` or `my-func(a, b, c)`)

why not the others?
- the .default suffix seems nice is a bit too long and doesn't really _fit_ with other terms defined alongside the default export
- plain `default` or `export=` is a bit too ambiguous
- using the module name verbatim is non-JS syntax if it has a dash in it (common)
- using the symbol name sometimes gets you unsavory implementation-detail results (like if `export default temp` or some other short name)

the caveat with messing with these default export names tho is that you've **potentially confused users** into thinking that it's a named export.

---

ideally it'd be nice to have something like this where the (default) is _there_ and informing you that, yes, this is the default export and not a named export BUT also being out of the way enough to not clutter it? idk the technical feasibility of this tho

<img width=300 src="https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/f8723972-49d0-4b08-ab1a-da1d2e723a6e">
<img width=300 src="https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/3ab646b4-7b73-403e-9af8-8c409403c4df">

you can _sorta_ do this now with the `@exports default` or some other `@summary THIS IS THE DEFAULT EXPORT` in your jsdoc comment but it's not quite the same lol
<img width=300 src="https://github.com/felipecrs/typedoc-plugin-rename-defaults/assets/61068799/ea95c7e7-5ddf-4f6e-b9ac-02c4d54a55d4">

```js
/**
 * @exports default
 * @name default
 * @summary THIS IS THE DEFAULT EXPORT
 * @defaultExport
 */
export default function assign<T extends object, U extends object>(o1: T, o2: U): T & U {
  return Object.assign(o1, o2)
}
```

**Comment by felipecrs** [comment]
We both agree on the module name camel cased option then, but here are two open points about it:

1. Although this is a common convention for JavaScript and TypeScript, some users might prefer for whatever reason a different style like `my_function`. I believe it is acceptable as an initial solution to do camel case and introduce more flexibility if there is demand.

2. When using the module name we stumble upon #4 again. I really don't know what is the best way to handle this, but it is indeed a valid use case (when using Angular and Nest conventions).

**Comment by felipecrs** [comment]
Maybe the best is just to keep current behavior of this plugin, except by camel casing the module name in the situations it is used.

> using the symbol name sometimes gets you unsavory implementation-detail results (like if export default temp or some other short name)

I am not very worried about this, because it is something that can be easily solved in the code itself (just give the symbol a better name).

**Comment by felipecrs** [comment]
About 2, maybe we can have a list of suffixes that would automatically get discarded from the module name. Which could also be extended with configuration.

```json
{
  "modulesSuffixesToDiscard": [
    "entity",
    "model",
    "service",
    "namespace",
    "provider"
  ]
}
```

**Comment by jcbhmr** [comment]
i have used `.test.ts` naming convention but not any of the `.entity.ts` or `.service.ts` etc that was mentioned.

imo since the `.{suffix}.ts` _seems_ (anecdotal) less common in general it makes sense to _default_ to using the file name verbatim where applicable and then encourage (in the readme or similar) using `/** @module withoutDotEntitySuffix */` to customize the module name for the `User.entity.ts` cases? that would seem the most extensible and least surprising solution idk.

or you could do the option thing with suffixes to auto-discard. 🤷‍♂️ either works; just make sure to add any magic conditions to the readme and you're good lol 😉

**Comment by felipecrs** [comment]
You are right, probably it's not the best to assume anything but instead have the feature of discard module suffixes based on the user's configuration.

While `/** @module withoutDotEntitySuffix */`, I would not like to push this as the solution because users can potentially have a lot of files and this would need to be set on a per-file basis.

**Comment by felipecrs** [comment]
But let me ask a real user for his opinion...

@AxeemHaider, since https://github.com/felipecrs/typedoc-plugin-rename-defaults/pull/4, modules named `User.entity.ts` gets named in TypeDoc as just `User`.

However, @jcbhmr found some interesting use cases where the module name would contain a dot and it would not be desired to discard part of the suffix. Instead, it would be desired to camel-case the module name.

If we apply this suggestion to everywhere, the symbol name for you, instead of just `User`, would become `UserEntity`.

Honestly it does not look too bad. I kinda like it. What do you think? It is a little hard to add do special handling for it otherwise, the alternative so far would be for the user to specify manually the modules suffixes they want to drop in the typedoc configuration.

**Comment by jcbhmr** [comment]
oh; i wanted to point this out: since the "use the symbol's name" happens _before_ falling back the module name, if you have a file like `User.service.ts` or something and you `export default class User {}` it will use the `User` name and not the `User.service` module name.

in essence, the only way you can get `my-wierd.name` is to **not** have a named symbol as the `export = myVar` or `export default myVar`

**Comment by felipecrs** [comment]
I think this situation is a little bit different. When I said:

> I am not very worried about this, because it is something that can be easily solved in the code itself (just give the symbol a better name).

I was referring to a situation where the symbol already had a name, although a bad one.

However, in the current situation, we are talking about when the symbol has no name at all, and refactoring the code to add names to such symbols can be cumbersome when having a lot of files.

**Comment by felipecrs** [comment]
Released as of v0.7.0.
